src/main/kotlin/com/chloe/kotlinserv/VertxHttpServer.kt

```diff
@@ -16,20 +17,47 @@ class VertxHttpServer : HttpServer {
     }
 
     private fun deployVertxRoute(route: HttpRoute, router: Router) {
+        router.route().handler(BodyHandler.create())
+
         if (route.method == HttpMethod.GET) {
             router.get(route.endpoint).handler { ctx ->
-                val httpResponse = route.processFunction.invoke()
+                val data = buildHttpResponse(ctx)
+                val httpResponse = route.processFunction.invoke(
+                    HttpRequest(requestHeaders = data[0], body = null, queryParameters = data[1])
+                )
                 convertHttpResponse(ctx, httpResponse)
             }
         } else {
             router.post(route.endpoint).handler { ctx ->
-                val httpResponse = route.processFunction.invoke()
+                val data = buildHttpResponse(ctx)
+                val body = ctx.body().asString()
+                val httpResponse = route.processFunction.invoke(
+                    HttpRequest(requestHeaders = data[0], body = body, queryParameters = data[1])
+                )
                 convertHttpResponse(ctx, httpResponse)
             }
         }
     }
 
     private fun convertHttpResponse(ctx: RoutingContext, httpResponse: HttpResponse) {
-        ctx.response().setStatusCode(httpResponse.code).end(httpResponse.responseBody)
+        if (httpResponse.responseBody == null) {
+            ctx.response().setStatusCode(httpResponse.code).end()
+        } else {
+            ctx.response().setStatusCode(httpResponse.code).end(httpResponse.responseBody)
+        }
+    }
+
+    private fun buildHttpResponse(ctx: RoutingContext): List<Map<String, List<String>>> {
```

Comment:
your returned value will be very difficult to understand by another developers and name of method don't match with result

Much better if you will just this part 
`.map { it.key to it.value }.groupBy({ it.first }, { it.second }).toMap()`
to separate method